### ISSUE TITLE
Last seen

### ISSUE THREAD

**Issue by simison** [Trustroots/trustroots#160]
- Save "last seen" date for:
  - stats, we need to understand better how people use it
  - can be used in sorting at search
- Show it at profile (and the API) in fuzzy format (?)

Lazy [Mongo update](http://blog.mongodb.org/post/248614779/fast-updates-with-mongodb-update-in-place) is great for this

Telegram [has nice set](https://telegram.org/faq/en#q-can-i-hide-my-last-seen-time) of fuzzy values:
- Last seen recently — covers anything between 1 second and 2-3 days
- Last seen within a week — between 2-3 and 7 days
- Last seen within a month — between 6-7 days and a month
- Last seen long time ago — more than a month (this is also always shown to blocked users)

In TR's case a mere month might be too short to slam "long time ago" on profiles. Half a year—1 year might be more suitable, following Warmshowers.org example.

**Comment by guaka** [comment]
Why hide it?

sent from my tablet

**Comment by simison** [comment]
I was thinking for user comfort (exact time-date feels a bit invading), but actually just making it fuzzy enough would help for that.

"Seen today" could be the most accurate.

**Comment by guaka** [comment]
Seen today
Seen yesterday
Seen last week
Seen last month
AWOL

**Comment by libertedeparole** [comment]
"last seen" by who?

**Comment by roberts-gh** [comment]
" 'last seen' by who?"

The server; could also be labeled "last login" like at CS.  Several popular txt message apps use the phrase "last seen," though in these cases, you're always logged in, so it just refers to any recent activity.

**Comment by libertedeparole** [comment]
My assumption is that this would be considered as an indication of the user being active or not. 
For that goal, I consider it a false "good idea". 

An "active" host in Berlin or Sund Finland will not receive the same amount of traffic. Some people will only log-in when the receive a request (or are traveling themselves). The Sund host might be very willing to host, but only receive a request every few years, whereas the Paris host will get them everyday, based on the facts on other Hospex sites.

If the laudable goal is to learn if the user is active or not, then I have other suggestions.

**Comment by libertedeparole** [comment]
IOW, this seems like imitating imperfect tools from other sites, giving a false sense of information.

**Comment by simison** [comment]
@libertedeparole nobody's going to look for a host from Paris and Sund _at the same time_. ;-)

**Comment by libertedeparole** [comment]
I guess my hitching trips are longer than yours.  :-))

I'm not sure how notification will develop at TR, both CS & BW require logging-in to access requests. But even if TR follows their model, last log-in still seems like a flimsy indicator to me. Is there another use to it besides indicating whether a user is "active" or not?

**Comment by simison** [comment]
I mean, I understood you're worried that at search results someone is pushed lower because of last seen timestamp due another one lives in the bushes and another one in the capital. Such search result you're never going to do. You will compare Lund hosts last logged times between each other, not to those in Paris.

I don't yet have idea how else to use this except to show it at the profile and leave it for people to decide how they interpret it instead of a machine. We can also use it for sorting search results, but for that member numbers won't be high enough for aaages.

Are you actually stating if someone logged in _today_ is not more likely to reply quick compared to someone from the same location who logged in, say, two years ago?

**Comment by simison** [comment]
On technical note;
- logging this only on node init at the backend should be enough
- If stats show up loads of people who leave the site open for longer than a day, some $interval for the frontend would be good but I really doubt this.

**Comment by guaka** [comment]
(It would be nice to include last seen through email activity.)

**Comment by simison** [comment]
>(It would be nice to include last seen through email activity.)
Valid point. blank.gif? :-)

**Comment by roberts-gh** [comment]
email image loads often blocked by default

**Comment by simison** [comment]
>email image loads often blocked by default

Yep I know. Better than nothing.

**Comment by libertedeparole** [comment]
Here’s how I see the question:

1) The feature is not a priority because for the moment “last log-in” does not correlate with “is this person a likely host?”, but only with “is this person an Internet junky?”. For now, the feature is not helpful. No one has not checked-in since 9 months ago. 

2) While “last sign-in” is a valid indicator of interest in TR, and (in the absence of forums) a fairly good indicator of willingness to host, it is not a good indicator of lack of interest in TR or in hosting.

3) The worst thing a site can do is remove a feature.

4) Mileage will differ in big cities, but based on BW traffic, for example, I am unlikely to get requests on TR more often than a couple times a year in Dijon (150,000 people and on a major transportation axe) for the next couple years. If I am notified of these requests by email, no need to check-in on TR between requests.

5) Hosting tends to enter into spirals. Let’s say for example, that TR has 10 hosts in Dijon. If two of those hosts find guests (for whatever reason), they become prime candidates for future requests, because they will have references/contacts. If this is reinforced by last sign-in date, they will receive more requests, simply because they are receiving more requests because they already hosted. This will make them access the site more often, adding to their “status” (If last login does have an effect on requests, the effect is obvious. If it has no effect, then the feature is useless.) This reinforces the spiral. Net effect: potential hosts are lost as requests focus on the top candidates. (There are of course other parallel effects that may undercut this spiral.)

6) I think it is possible to create a better indicator of inactive hosts in the future. This is not a priority for the moment, because of 1) above. This better indicator would remove the need for last log-in (see 3) above). 

7) The indicator of inactive hosts needs brainstorming, but I think last login, with the addition of query emails (“you haven’t signed on in 12 months, are you still interested in hosting?”) could effectively cull the membership rolls of non-hosters. (There are other metrics that could also be explored - would you like to host a) more b) less c) about the same - that could be exploited in different ways depending how the site is implemented.)(I don't mean to exclude guesters)

8) “Are you actually stating if someone logged in today is not more likely to reply quick compared to someone from the same location who logged in, say, two years ago?” 

Statistically, you are right. Or rather you will be right in the future. For now, it is dubious.

The question is: should we eliminate potential hosts by favoring “Internet junkies” and risking entering into the sort of spiral that is quite visible in a large system like CS? The moment there are a dozen hosts in a particular city, the spiral effect will begin, and increase as membership increases. An indicator such as last log-in will only increase those spirals.

Sorry for being so lengthy, but hopefully I have mentioned a few elements that also enter into other questions.

David

**Comment by globestoppeuse** [comment]
+1 to all that.

Activity indicator in the future could combine metrics about replied to
messages, time between message and reply and general website activity. A
renewal to the hosting listing seems fair after 6 months if there is no
activity. I think WS has a 1 year renewal link (and an invitatino to then
update your profile, if you forgot about it).

**Comment by simison** [comment]
Guys...

>For now, it is dubious.

This feature sits in "long term future" milestone.

**Comment by mrkvon** [comment]
A very basic version is implemented by #526.

## Todo

- [ ] show not exact, but fuzzy information in user profile
- [ ] use in statistics (perhaps daily statistics: how many users logged last day, week, month, year, ... never), (_other ideas?_)
- [ ] ...
